### PR TITLE
[DROOLS-7608] fix KieProject lookup on classpath for EAP 7.4.15 VFS

### PR DESCRIPTION
**Pull request by mariofusco**
@tkobayas This is only a rewritten and cleaned up version of the solution that you found [here](https://github.com/kiegroup/drools/pull/25).

I dug into VFS sources a lot and couldn't find any better solution. I understand that using reflection is not ideal, but in the end we were already doing so, even if just because we didn't want a direct dependency on VFS. Now it's true that this is a bit worse than before because we're also accessing package private methods (that are probably subject to change in future), but other than asking the Jboss team to make those methods public, thus avoiding to use `setAccessible(true)`, I don't see any possible improvement.

### CODE REVIEW

File: drools-compiler/src/main/java/org/drools/compiler/kie/builder/impl/ClasspathKieProject.java [hunk 1]
```diff
@@ -440,41 +440,50 @@ public static String fixURLFromKProjectPath(URL url) {
...
             } catch (Exception e1) {
                 // VirtualFile is not available on the classpath - ignore
                 log.warn( "Found virtual file " + url + " but org.jboss.vfs.VirtualFile is not available on the classpath" );
             }
         }
-        Method m2 = null;
+        
+        Class vfsClass = null;
+        Method vfsGetChildMethod = null;
+        boolean useTccl = false;
```
Comment (mariofusco) [inline drools-compiler/src/main/java/org/drools/compiler/kie/builder/impl/ClasspathKieProject.java:458]: @tkobayas this is the only relevant difference with your solution: we need to know from which classloader we loaded the VFS classes and always use the same consistently. I believe this is necessary for OSGi.

**Comment by gmunozfe** [comment]
Failures in the PR pipeline (drools-retediagram):

```
[2024-02-15T18:00:25.164Z]    [report] Classes in bundle 'drools-retediagram' do no match with execution data. For report generation the same class files must be used as at runtime.
[2024-02-15T18:00:25.164Z]    [report] Execution data for class org/drools/ecj/EclipseCompilationProblem does not match.
```

```
[ERROR] Failed to execute goal org.sonarsource.scanner.maven:sonar-maven-plugin:3.6.0.1398:sonar (default) on project drools-retediagram: 
Could not find the pullrequest with key '26': 
Error 404 on https://sonarcloud.io/api/alm_integration/show_pullrequest?project=org.drools%3Adrools&pullrequestKey=26 : 
{"errors":[{"msg":"The pullrequest could not be found"}]} -> [Help 1]
```

jenkins tests this

**Comment by gmunozfe** [comment]
@mariofusco @tkobayas I also patched kie-server war with this fix and when testing with EAP 7.14.5, found this issue:

_22:57:43.790 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor - [ERROR] Failed to execute goal org.kie:kie-maven-plugin:7.75.0-SNAPSHOT:build (default-build) on project scorecard_mm: Execution default-build of goal org.kie:kie-maven-plugin:7.75.0-SNAPSHOT:build failed: An API incompatibility was encountered while executing org.kie:kie-maven-plugin:7.75.0-SNAPSHOT:build: java.lang.NoSuchMethodError: 'boolean org.drools.compiler.compiler.AnalysisResult.isTypesafe()'_

_22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor - Caused by: java.lang.NoSuchMethodError: 'boolean org.drools.compiler.compiler.AnalysisResult.isTypesafe()'
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.mvel.builder.MVELDialect.getMVELCompilationUnit (MVELDialect.java:663)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.mvel.MVELConstraintBuilder.buildCompilationUnit (MVELConstraintBuilder.java:420)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.mvel.MVELConstraintBuilder.buildCompilationUnit (MVELConstraintBuilder.java:386)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.mvel.MVELConstraintBuilder.buildVariableConstraint (MVELConstraintBuilder.java:190)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.PatternBuilder.buildConstraintForPattern (PatternBuilder.java:1176)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.PatternBuilder.buildRelationalExpression (PatternBuilder.java:1037)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.PatternBuilder.buildExpression (PatternBuilder.java:907)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.PatternBuilder.buildCcdDescr (PatternBuilder.java:879)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.PatternBuilder.build (PatternBuilder.java:696)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.PatternBuilder.processConstraintsAndBinds (PatternBuilder.java:600)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.PatternBuilder.build (PatternBuilder.java:173)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.PatternBuilder.build (PatternBuilder.java:140)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.PatternBuilder.build (PatternBuilder.java:120)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.GroupElementBuilder.build (GroupElementBuilder.java:66)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.GroupElementBuilder.build (GroupElementBuilder.java:36)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.GroupElementBuilder.build (GroupElementBuilder.java:66)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.GroupElementBuilder.build (GroupElementBuilder.java:36)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.GroupElementBuilder.build (GroupElementBuilder.java:66)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.rule.builder.RuleBuilder.build (RuleBuilder.java:106)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.builder.impl.KnowledgeBuilderImpl.internalAddRule (KnowledgeBuilderImpl.java:1113)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.builder.impl.KnowledgeBuilderImpl.addRule (KnowledgeBuilderImpl.java:1108)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.builder.impl.KnowledgeBuilderImpl.compileAllQueries (KnowledgeBuilderImpl.java:1043)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.builder.impl.KnowledgeBuilderImpl.compileRules (KnowledgeBuilderImpl.java:1031)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.builder.impl.KnowledgeBuilderImpl.compileKnowledgePackages (KnowledgeBuilderImpl.java:852)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.builder.impl.KnowledgeBuilderImpl.buildRules (KnowledgeBuilderImpl.java:2420)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.builder.impl.KnowledgeBuilderImpl.doFirstBuildStep (KnowledgeBuilderImpl.java:2345)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build (CompositeKnowledgeBuilderImpl.java:109)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.builder.impl.CompositeKnowledgeBuilderImpl.build (CompositeKnowledgeBuilderImpl.java:99)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages (AbstractKieProject.java:270)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.kie.builder.impl.AbstractKieProject.buildKnowledgePackages (AbstractKieProject.java:216)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.kie.builder.impl.AbstractKieModule.buildKnowledgePackages (AbstractKieModule.java:193)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.kie.builder.impl.AbstractKieModule.createKieBase (AbstractKieModule.java:205)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.kie.builder.impl.KieContainerImpl.createKieBase (KieContainerImpl.java:475)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.drools.compiler.kie.builder.impl.KieContainerImpl.getKieBase (KieContainerImpl.java:439)
22:57:43.846 [pool-7-thread-1] INFO  o.k.s.i.shared.ProcessExecutor -     at org.kie.pmml.pmml_4_2.PMML4Compiler.checkBuildingResources (PMML4Compiler.java:441)_

**Comment by gmunozfe** [comment]
@mariofusco base is not updated, you have to rebase with 7.x

**Comment by gmunozfe** [comment]
jenkins do cdb

**Comment by tkobayas** [comment]
Filed a PR with the rebased one: https://github.com/kiegroup/drools/pull/27

You can close this PR.

**Review by tkobayas — APPROVED** [review]
Thanks @mariofusco , LGTM. Approved, but it's better to move to https://github.com/kiegroup/drools/pull/27 for local test convenience.

**Comment by mariofusco** [comment]
Replaced by https://github.com/kiegroup/drools/pull/27

**Pull request closed without merging**